### ISSUE TITLE
Precondition bug missed unless invoked within function with its own precondition

### ISSUE THREAD

**Issue by poneill** [pschanely/CrossHair#113]
**Is your feature request related to a problem? Please describe.**

Consider the following snippet:
```
# a simple function whose only job is to assert a precondition.
def f(x):
    """
    pre: x > 0
    """
    return None

# this does not cause an error, even though it seems it should
f(0)

# neither does this
def h1(x):
    return f(0)

# but if we add a trivial precondition, we're able to trigger it
def h2(x):
    """
    pre: x == x
    """
    return f(0)
```

**Describe the solution you'd like**
I think the modal user would expect the precondition violation to be caught in all three cases.  Failing that, again, it would probably help to have expanded documentation about crosshair's limitations.

**Describe alternatives you've considered**
N/A

Thanks!

**Comment by pschanely** [comment]
Ah. You aren't the first person to be led astray here. I've started with a documentation "fix."

Note also issue #22, which is highly related.

More than anything else, the current behavior is erring on the side of safety: we don't run code unless we're pretty sure you've declared it to be safe to run. (where "safe" means it won't write to disk, make network calls, consume infinite memory, etc).

I am very open to ideas about how to catch examples like yours without adding a "footgun." (and/or suggestions for how to talk about this behavior/problem in a clear way) I'll leave this bug open to collect further thoughts and ideas!

**Comment by poneill** [comment]
Sorry, I probably just don't understand enough about crosshair's internals to fully appreciate the problem here.  Is there documentation about how to declare a function safe / pure, or is the idea more that attaching any contract to a function tells crosshair that it's fair game to run?

**Comment by pschanely** [comment]
That's right; the way to tell CrossHair it's safe to run is to attach a contract to it. (at least right now)
Discussion #107 is also relevant.

An aside: I feel like there might be value in a convention or tiny 3rd party library that allows functions to declare themselves "pure."  A non-CrossHair example: Nukia could more easily perform [certain optimizations](https://github.com/Nuitka/Nuitka#quicker-function-calls) where evaluation order is known to be alterable.
